Sources/Common/Util/ThreadUtil.swift

```diff
@@ -1,18 +1,27 @@
 import Foundation
 
-// allows us to more easily have automated tests with threading
 public protocol ThreadUtil {
-    func runBackground(_ block: @escaping () -> Void)
-    func runMain(_ block: @escaping () -> Void)
+    // It's important that these functions work as a FIFO serial queue. Our code depends on the blocks of code being executed in order, not concurrently.
+    func queueOnBackground(_ block: @escaping () -> Void)
+    func queueOnMain(_ block: @escaping () -> Void)
 }
 
+// Benefits of this object:
+// 1. Allows us to mock threading behavior in unit tests if appropriate.
+// 2. Our SDK can have a singleton queue that runs blocks of code on a background thread in FIFO order.
+//
+// Important: This class must be a singleton so only 1 queue object is used by the SDK code.
+//
 // sourcery: InjectRegister = "ThreadUtil"
+// sourcery: InjectSingleton
 public class CioThreadUtil: ThreadUtil {
-    public func runMain(_ block: @escaping () -> Void) {
+    private let sharedBackgroundQueue = DispatchQueue(label: "io.customer.sdk.shared_background", qos: .background)
+
+    public func queueOnMain(_ block: @escaping () -> Void) {
         DispatchQueue.main.async(execute: block)
     }
 
-    public func runBackground(_ block: @escaping () -> Void) {
-        DispatchQueue.global(qos: .background).async(execute: block)
+    public func queueOnBackground(_ block: @escaping () -> Void) {
+        sharedBackgroundQueue.async(execute: block)
```

Comment:
was this the root of the problem that we were using `DispatchQueue.global`?
if so could use a comment as to why using `sharedBackgroundQueue` moves off the main thread.

Reply:
The root problem was we were not using any `DispatchQueue`, we were writing files from the thread that the caller called `CustomerIO.X` function on. 

[These comments in the file](https://github.com/customerio/customerio-ios/blob/b1fd8ff69c03ff3aaa98ab2f0b8bda2af39f71b9/Sources/Common/Util/ThreadUtil.swift#L11-L13) are my attempt at explaining the use of `sharedBackgroundQueue` in the file. If those comments don't explain as well as you hope, comment with a suggestion.